### PR TITLE
Gjp/browser fixes

### PR DESCRIPTION
**Pull request by gpittarelli**
Fixes for various weird behavior we've noticed after rolling out my previous PRs:
- `visitId` -> `visit` typo (didn't break anything except interrupted sessions restoring to the right scroll position)
- NEVER access `sessionStorage` without try/catch. Some browsers will throw exceptions for simply accessing the object (which, in my strict reading of the spec seems nonconforming). Also closes the edge case of a legitimate QuotaExceeded exception occurring between our feature test and not happening when we actually generate the visitId.
- make `stateIds.isBefore` safe for undefined values
- use pageshow/pagehide instead of beforeunload, when available ( https://webkit.org/blog/516/webkit-page-cache-ii-the-unload-event/ )

Testing: See same steps from this PR: https://github.com/Massdrop/fluxible-router/pull/8
Except: in the middle of the 2nd and 3rd pages: visit an external page (type it in URL bar) and then return to the test site by manually typing the url (so you have a session history of: our site p1, our site p2, external site, our site p3)

---
This change is [Reviewable](https://reviewable.io/reviews/massdrop/fluxible-router/12)

### CODE REVIEW

File: lib/stateIds.js [hunk 1]
```diff
@@ -29,24 +29,15 @@ var nextStateCounter = (function () {
...
-    if (hasSessionStorage) {
+    try {
         if (window.sessionStorage.fluxibleRouterStateId) {
             nextId = ++window.sessionStorage.fluxibleRouterStateId;
         } else {
             nextId = window.sessionStorage.fluxibleRouterStateId = 1;
         }
-    } else {
+    } catch (e) {
         nextId = +(new Date());
```
Comment (geuis) [inline lib/stateIds.js]: Don't need to do this. Just use Date.now(). And if you need to do type conversion, generally its better to avoid typecasting tricks where possible. They're usually just as performant, but aren't as clear as to what's happening when others are reading through the code.

File: lib/handleHistory.js [hunk 2]
```diff
@@ -228,7 +233,13 @@ function createComponent(Component, opts) {
 
         componentWillUnmount: function () {
             this._history.off(this._onHistoryChange);
-            window.removeEventListener('beforeunload', this.beforeunload);
+
+            if ('onpagehide' in window) {
```
Comment (geuis) [inline lib/handleHistory.js]: indentation should be 2 chars

File: lib/stateIds.js [hunk 3]
```diff
@@ -29,24 +29,15 @@ var nextStateCounter = (function () {
...
 
-    if (hasSessionStorage) {
+    try {
         if (window.sessionStorage.fluxibleRouterStateId) {
             nextId = ++window.sessionStorage.fluxibleRouterStateId;
         } else {
             nextId = window.sessionStorage.fluxibleRouterStateId = 1;
         }
-    } else {
+    } catch (e) {
```
Comment (gpittarelli) [inline lib/stateIds.js]: Leaving out the `e` there is a SyntaxError, and eslint doesn't complain about it. (check http://eslint.org/docs/rules/no-unused-vars they specifically default to silencing unused catch variables)

File: .eslintrc [hunk 4]
```diff
@@ -6,7 +6,7 @@
     strict: 0
     indent: [2, 4, {SwitchCase: 1}]
     quotes: [2, 'single']
-    no-unused-vars: 0 // see https://github.com/babel/babel-eslint/issues/21
+    no-unused-vars: 2
```
Comment (geuis) [inline .eslintrc]: Kind of a nuke from orbit solution here. See the next comment about how to disable this one a line by line basis.

File: lib/stateIds.js [hunk 5]
```diff
@@ -29,25 +29,16 @@ var nextStateCounter = (function () {
...
-    if (hasSessionStorage) {
+    try {
         if (window.sessionStorage.fluxibleRouterStateId) {
             nextId = ++window.sessionStorage.fluxibleRouterStateId;
         } else {
             nextId = window.sessionStorage.fluxibleRouterStateId = 1;
         }
-    } else {
-        nextId = +(new Date());
+    } catch (e) {
```
Comment (geuis) [inline lib/stateIds.js]: Can add `//eslint-disable-line` to prevent eslint from checking this
    try {
     ...
    } catch (e) { //eslint-disable-line
        nextId = Date.now();
    }

Comment (gpittarelli) [inline lib/stateIds.js]: It already doesn't check that

File: .eslintrc (hunk 4 above)
Comment (gpittarelli) [inline .eslintrc]: no-unused-vars was only disabled because of the linked issue in babel-eslint, which is closed now, so its safe to reenable... and might as well
